### PR TITLE
fix(scanner-v4-db): Fix scanner-v4-db init

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Scanner V4 DB needs to be initialized, current state when starting the `quay.io/rhacs-eng/scanner-v4-db` image:

```
Error: Database is uninitialized and superuser password is not specified.
       You must specify POSTGRES_PASSWORD to a non-empty value for the
       superuser. For example, "-e POSTGRES_PASSWORD=password" on "docker run".

       You may also use "POSTGRES_HOST_AUTH_METHOD=trust" to allow all
       connections without a password. This is *not* recommended.

       See PostgreSQL documentation about "trust":
       https://www.postgresql.org/docs/current/auth-trust.html
```
An `init-entrypoint.sh` script is ran by an init container to initialize `central-db`, the same script is now used to initialize `scanner-v4-db`.

Additionally, the scanner `Makefile` assumed `GOARCH` was exported, which if not set in host environment would cause `make images` / `make db-image` to fail with: 

```
$ cd scanner

$ make db-image
+ copy-db-scripts
+ copy-db-sigs
+ db-image
Building scanner-db:4.2.x-205-gd7925a54b6-dirty
Building with platform linux/
  buildx: Docker Buildx (Docker Inc.)
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
ERROR: "" is an invalid component of "linux/": platform specifier component must match "^[A-Za-z0-9_-]+$": invalid argument
make: *** [Makefile:104: db-image] Error 1
```
The `make/env.mk` script used by the main ACS Makefile exports `GOARCH` (amongst other things) and was re-used here.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [x] Determined and documented upgrade steps
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

### Here I tell how I validated my change

In a linux VM, built the scanner-v4 images and pushed db image to quay.  Then created the `scanner-v4-db` deployment using the WIP helm charts that includes an `initContainer` attempting to execute `init-entrypoint.sh`

```
cd scanner
make images
docker tag stackrox/scanner-v4-db:4.2.x-205-gd7925a54b6-dirty quay.io/dcaravel/scanner-v4-db:latest
docker push
```
#### Before Fix
```
$ k describe pod scanner-v4-db-847f5d686c-nj9wz
...
  Warning  Failed          13s                kubelet            Error: container create failed: time="2023-09-20T22:43:37Z" level=error msg="runc create failed: unable to start container process: exec: \"init-entrypoint.sh\": executable file not found in $PATH"
```

#### After Fix
`scanner-v4-db` now starts
##### `init-db` container logs
```
The files belonging to this database system will be owned by user "postgres".
This user must also own the server process.

The database cluster will be initialized with locale "en_US.utf8".
The default database encoding has accordingly been set to "UTF8".
The default text search configuration will be set to "english".

Data page checksums are enabled.

fixing permissions on existing directory /var/lib/postgresql/data/pgdata ... ok
creating subdirectories ... ok
selecting dynamic shared memory implementation ... posix
selecting default max_connections ... 100
selecting default shared_buffers ... 128MB
selecting default time zone ... UTC
creating configuration files ... ok
running bootstrap script ... ok
performing post-bootstrap initialization ... ok
syncing data to disk ... ok

Success. You can now start the database server using:

    pg_ctl -D /var/lib/postgresql/data/pgdata -l logfile start
```
##### `db` container logs
```
PostgreSQL Database directory appears to contain a database; Skipping initialization

Waiting for child process 14 to exit
2023-09-20 22:27:59.167 UTC [15] LOG:  starting PostgreSQL 13.12 on x86_64-pc-linux-gnu, compiled by gcc (GCC) 8.5.0 20210514 (Red Hat 8.5.0-18), 64-bit
2023-09-20 22:27:59.168 UTC [15] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-09-20 22:27:59.168 UTC [15] LOG:  listening on IPv6 address "::", port 5432
2023-09-20 22:27:59.171 UTC [15] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-09-20 22:27:59.175 UTC [15] LOG:  listening on Unix socket "/tmp/.s.PGSQL.5432"
2023-09-20 22:27:59.184 UTC [16] LOG:  database system was shut down at 2023-09-20 22:27:57 UTC
2023-09-20 22:27:59.191 UTC [15] LOG:  database system is ready to accept connections
```

### Reminder for reviewers

In addition to reviewing code here, reviewers **must** also review testing and request further testing in case the
performed one does not seem sufficient. As a reviewer, you must not approve the change until you understand the
performed testing and you are satisfied with it.

### CODE REVIEW

File: scanner/Makefile
Comment (jvdm) [reply scanner/Makefile:37]: There is `export GOBIN`, which is not necessary anymore since we are using `env.mk`.

Comment (jvdm) [reply scanner/Makefile:37]: Shouldn't we just actually export GOARCH in `scanner/Makefile` for simplicity?

**Review by jvdm — COMMENTED** [review]
I am wondering if we should use:

```
DOCKERBUILD := GOARCH=$(GOARCH) $(CURDIR)/../scripts/docker-build.sh
```

And avoid setting GOARCH everywhere and change behaviour when building the build tools that should run in the build architecture which can be different from the target architecture?

File: scanner/Makefile
Comment (dcaravel) [reply scanner/Makefile:34]: Indeed, I saw there were a bunch of things in `env.mk` that would replace whats currently in the `scanner/Makefile` - I started cleaning it up but it was turning into a slippery slope trying to do it 'cleanly' - I opted for the quick fix. I've done a bit more cleanup in https://github.com/stackrox/stackrox/pull/7844/commits/04e690ebc24f005b691768b7cd7103aae0da1329

File: scanner/image/db/scripts/init-entrypoint.sh
Comment (dcaravel) [reply scanner/image/db/scripts/init-entrypoint.sh:1]: Interesting `Makefile` wizardry there, the resulting command ended up being:
```
cp ../image/postgres/download.sh ../image/postgres/scripts/init-entrypoint.sh image/db
```
Wasn't clear to me how to change the `cp` command to choose a different destination dir based on the input, so I created another target instead. https://github.com/stackrox/stackrox/pull/7844/commits/04e690ebc24f005b691768b7cd7103aae0da1329 and https://github.com/stackrox/stackrox/pull/7844/commits/f80de7d599b84404868fe8eaafad16dc4e0398f4

**Comment by dcaravel** [comment]
> I am wondering if we should use:
> 
> ```
> DOCKERBUILD := GOARCH=$(GOARCH) $(CURDIR)/../scripts/docker-build.sh
> ```
> 
> And avoid setting GOARCH everywhere and change behaviour when building the build tools that should run in the build architecture which can be different from the target architecture?

Similarly I was wondering why not using a 'builder image' for this - I had to hack up an install of a different version of `go` to get past the `EXPECTED_GO_VERSION` check.  Perhaps exporting the `GOARCH` is a non-issue if using a builder image?

**Comment by jvdm** [comment]
> > I am wondering if we should use:
> > ```
> > DOCKERBUILD := GOARCH=$(GOARCH) $(CURDIR)/../scripts/docker-build.sh
> > ```
> > And avoid setting GOARCH everywhere and change behaviour when building the build tools that should run in the build architecture which can be different from the target architecture?
> 
> Similarly I was wondering why not using a 'builder image' for this - I had to hack up an install of a different version of `go` to get past the `EXPECTED_GO_VERSION` check. Perhaps exporting the `GOARCH` is a non-issue if using a builder image?

For reproducible builds, yes, containers. But I am also OK supporting dirty local builds for faster development cycles (debugging, local runs, etc.).  The benefit of faster cycles during dev. are greater than sometimes the patch being blocked by CI -- which is supposed to run in containers and prevent wrong assumptions in your local from affecting the repo.

File: scanner/Makefile
Comment (jvdm) [reply scanner/Makefile:34]: Yes, there might be other things. It's OK for now. Thanks for improving it.

File: scanner/Makefile
Comment (jvdm) [inline scanner/Makefile:145]: Nitpick: I haven't realized you wanted to keep the same script structure. We could put everything under `scripts/`, for example (like moving the ref to `scripts/download.sh` in the Dockerfile). Nonetheless, if you want to keep the original structure, this could work:

```suggestion
.PHONY: copy-db-scripts

# Files to copy.
copy-db-scripts := scripts/init-entrypoint.sh download.sh

# Directories.
copy-db-src     := ../image/postgres/
copy-db-dest    := image/db/

copy-db-scripts: $(addprefix $(copy-db-src),$(copy-db-scripts))
	$(SILENT)echo "+ $@"
	$(SILENT)tar -C $(copy-db-src) -cf - $(patsubst $(copy-db-src)%,./%,$?) \
	    | tar -C $(copy-db-dest) -xf -
```

Not tested since I am just nitpicking. It uses `tar` for compatibility, `cp` has `--parents`, but that would not run outside GNU/Linux.

**Review by jvdm — APPROVED** [review]
LGTM. We can follow up on env and container builds later on.

**Comment by dcaravel** [comment]
> For reproducible builds, yes, containers. But I am also OK supporting dirty local builds for faster development cycles (debugging, local runs, etc.). The benefit of faster cycles during dev. are greater than sometimes the patch being blocked by CI -- which is supposed to run in containers and prevent wrong assumptions in your local from affecting the repo.

Agreed with faster cycles, I wasn't referring to CI - was referring to performing local builds of the binary in a container (like the existing ACS builds).  The  'extra environment setup' that has to be done to execute the local image builds (ie: must have `go 1.20.4` exactly) shouldn't be necessary.

File: scanner/Makefile
Comment (dcaravel) [reply scanner/Makefile:145]: >  haven't realized you wanted to keep the same script structure
I'm indifferent on changing the structure, I just assumed keep the existing structure used by `central-db` and `scanner-db` today to not 'rock the boat' unnecessarily.

**Review by RTann — APPROVED** [review]
Approving, as it all makes sense. See if you want to consider Victor's suggestion before merging

**Comment by RTann** [comment]
> Similarly I was wondering why not using a 'builder image' for this - I had to hack up an install of a different version of `go` to get past the `EXPECTED_GO_VERSION` check. Perhaps exporting the `GOARCH` is a non-issue if using a builder image?

I think there were multi-arch build reasons behind removing this, but I can't remember anymore. Maybe @connorgorman recalls?

**Pull request merged**
